core/shared/src/main/scala/zio/internal/ConcurrentWeakHashSet.scala

```diff
@@ -0,0 +1,579 @@
+   * calculate the size of the segments and the size of the reference array.
+   */
+  private[internal] def calculateShift(minimumValue: Int, maximumValue: Int): Int = {
+    var shift = 0
+    var value = 1
+    while (value < minimumValue && value < maximumValue) {
+      value <<= 1
+      shift += 1
+    }
+    shift
+  }
+
+}
+
+/**
+ * A `Set` data type that uses weak references to store its elements in highly
+ * concurrent environment. This is faster alternative to
+ * `Collections.synchronizedSet(Collections.newSetFromMap(new WeakHashMapA,
+ * java.lang.Boolean))`.
+ *
+ * Implementation of this class is a combination of some existing solutions:
+ *   - `ConcurrentHashMap` from JDK 7 - for concurrency control using segments
+ *   - `WeakHashMap` - for storing weak references to elements
+ *   - `ConcurrentReferenceHashMap` from Spring Framework - the reference map
+ *     that combines references & with concurrent segments
+ *
+ * This implementation offers better performance than synchronized set for
+ * concurrent operations, but might be a little bit slower for serial
+ * (single-threaded only) access.
+ *
+ * @param initialCapacity
+ *   initial size of the set (default
+ *   [[ConcurrentWeakHashSet.DefaultInitialCapacity]])
+ * @param loadFactor
+ *   load factor for the set that defines how fast it will grow (default
+ *   [[ConcurrentWeakHashSet.DefaultLoadFactor]])
+ * @param concurrencyLevel
+ *   approximate number of threads that will access the set concurrently
+ *   (default [[ConcurrentWeakHashSet.DefaultConcurrencyLevel]])
+ * @tparam V
+ *   type of the elements stored in the set
+ */
+private[zio] class ConcurrentWeakHashSet[V](
+  initialCapacity: Int = ConcurrentWeakHashSet.DefaultInitialCapacity,
+  loadFactor: Float = ConcurrentWeakHashSet.DefaultLoadFactor,
+  concurrencyLevel: Int = ConcurrentWeakHashSet.DefaultConcurrencyLevel
+) extends mutable.Set[V] { self =>
+
+  private val shift =
+    ConcurrentWeakHashSet.calculateShift(this.concurrencyLevel, ConcurrentWeakHashSet.MaxConcurrencyLevel)
+  private var segments: Array[Segment] = _
+
+  {
+    val size                     = 1 << shift
+    val roundedUpSegmentCapacity = ((this.initialCapacity + size - 1L) / size).toInt
+    val initialSize =
+      1 << ConcurrentWeakHashSet.calculateShift(roundedUpSegmentCapacity, ConcurrentWeakHashSet.MaxSegmentSize)
+    val segments        = new Array[Segment](size)
+    val resizeThreshold = (initialSize * this.loadFactor).toInt
+    segments.indices.foreach(i => segments(i) = new Segment(initialSize, resizeThreshold))
+    this.segments = segments
+  }
+
+  /**
+   * Enforce cleanup of dead references in the set. By default cleanup is
+   * performed automatically when the set is modified, but some operations (like
+   * e.g. `size`) do not modify the set and therefore do not trigger cleanup.
+   */
+  def gc(): Unit =
+    this.segments.foreach(segment => segment.restructureIfNecessary(false))
```

Comment:
Rewrite to use a `while` loop and index to eliminate the allocation of the closure.